src/lib/log/src/lib.rs

```diff
@@ -94,16 +94,21 @@ macro_rules! println {
 /// Error types that may happen with serial transfer, usually just `WouldBlock`
 #[derive(Debug)]
 pub struct Error {
-    pub kind: embedded_hal::serial::ErrorKind,
+    pub kind: embedded_hal_nb::serial::ErrorKind,
 }
 
-impl embedded_hal::serial::Error for Error {
-    fn kind(&self) -> embedded_hal::serial::ErrorKind {
+impl embedded_hal_nb::serial::Error for Error {
+    fn kind(&self) -> embedded_hal_nb::serial::ErrorKind {
         self.kind
     }
 }
 
+#[cfg(not(feature = "jh71xx"))]
```

Comment:
Is there a chance that we could avoid specific implementations leaking into this library?

Reply:
We could, but it would require each implementation to have a HAL that implements the `embedded_hal_nb::serial` traits.

The reason I had to rework the trait here is because `jh71xx-hal` already implements these traits. The other `mainboard` implementations do not. I can try to rework things here again, and see if I can avoid conditional compilation. I agree that would be preferable.

Reply:
We could, but it would require each implementation to have a HAL that implements the `embedded_hal_nb::serial` traits.

The reason I had to rework the trait here is because `jh71xx-hal` already implements these traits. The other `mainboard` implementations do not. I can try to rework things here again, and see if I can avoid conditional compilation.

Implementations should be moved into each `mainboard` SoC implementation. There would still need to be conditional compilation based on which `mainboard` is being compiled, but could use generic type arguments to hide implementation details.

It's also because you define the type alias `SerialLogger` with the concrete `log::Error` type. Then you store a trait object as a mutable static, which prevents supplying a generic argument if we swapped the concrete `log::Error` for a generic type argument.

The design pattern I've seen in most HALs is that an `Error` type is defined per-module implementing each HAL trait.

As a shim placeholder until each `mainboard` SoC has a HAL crate, the `Serial` and `embedded_hal_nb::serial` traits should be implemented for each `mainboard` SoC UART peripherals. Then, we could do single conditional compilation based on the implementation from each module.

Reply:
Alright, I added a wrapper type in `visionfive2/bt0/src/uart` so we can impl the `log::Serial` trait there.